ci/jenkins/test-rancher.sh

```diff
@@ -0,0 +1,387 @@
+    echoerr "$_usage"
+}
+
+function print_help {
+    echoerr "Try '$0 --help' for more information."
+}
+
+while [[ $# -gt 0 ]]
+do
+key="$1"
+
+case $key in
+    --cluster-name)
+    CLUSTER_NAME="$2"
+    shift 2
+    ;;
+    --workdir)
+    WORKDIR="$2"
+    shift 2
+    ;;
+    --testcase)
+    TESTCASE="$2"
+    shift 2
+    ;;
+    --registry)
+    DOCKER_REGISTRY="$2"
+    shift 2
+    ;;
+    --testbed-type)
+    TESTBED_TYPE="$2"
+    shift 2
+    ;;
+    --proxyall)
+    PROXY_ALL=true
+    shift
+    ;;
+    --ip-mode)
+    IP_MODE="$2"
+    shift 2
+    ;;
+    -h|--help)
+    print_usage
+    exit 0
+    ;;
+    *)    # unknown option
+    echoerr "Unknown option $1"
+    exit 1
+    ;;
+esac
+done
+
+if [[ "${IP_MODE}" == "" ]]; then
+    IP_MODE=${DEFAULT_IP_MODE}
+fi
+if [[ "${IP_MODE}" != "${DEFAULT_IP_MODE}" && "${IP_MODE}" != "ipv6" && "${IP_MODE}" != "ds" ]]; then
+    echoerr "--ip-mode must be ipv4, ipv6 or ds"
+    exit 1
+fi
+if [[ "$WORKDIR" != "$DEFAULT_WORKDIR" && "$KUBECONFIG_PATH" == "$DEFAULT_KUBECONFIG_PATH" ]]; then
+    KUBECONFIG_PATH=${WORKDIR}/.kube/config
+fi
+NO_PULL=
+if [[ "$DOCKER_REGISTRY" != "" ]]; then
+    # Image pulling policy of Sonobuoy is 'Always' by default. With dockerhub rate limit, sometimes it is better to use
+    # cache when registry is used.
+    IMAGE_PULL_POLICY="IfNotPresent"
+    # If DOCKER_REGISTRY is non null, we ensure that "make" commands never pull from docker.io.
+    NO_PULL=1
+fi
+export NO_PULL
+
+E2ETEST_PATH=${WORKDIR}/kubernetes/_output/dockerized/bin/linux/amd64/e2e.test
+
+function fetch_kubeconfig {
+    rancher cluster kf $CLUSTER_NAME -o yaml >> $KUBECONFIG_PATH
+}
+
+function export_govc_env_var {
+    export GOVC_URL=$GOVC_URL
+    export GOVC_USERNAME=$GOVC_USERNAME
+    export GOVC_PASSWORD=$GOVC_PASSWORD
+    export GOVC_INSECURE=1
+    export GOVC_DATACENTER=$GOVC_DATACENTER
+    export GOVC_DATASTORE=$GOVC_DATASTORE
+}
+
+function clean_antrea {
+    echo "====== Cleanup Antrea Installation ======"
+    clean_up_one_ns "monitoring"
+    clean_up_one_ns "antrea-ipam-test-11"
+    clean_up_one_ns "antrea-ipam-test-12"
+    clean_up_one_ns "antrea-ipam-test"
+    clean_up_one_ns "antrea-test"
+    # Delete antrea-prometheus first for k8s>=1.22 to avoid Pod stuck in Terminating state.
+    kubectl delete -f ${WORKDIR}/antrea-prometheus.yml --ignore-not-found=true || true
+    for antrea_yml in ${WORKDIR}/*.yml; do
+        kubectl delete -f $antrea_yml --ignore-not-found=true || true
+    done
+    docker images | grep 'antrea' | awk '{print $3}' | xargs -r docker rmi || true
+    docker images | grep '<none>' | awk '{print $3}' | xargs -r docker rmi || true
+}
+
+function clean_vm_agent {
+    echo "====== Cleanup Antrea-Agent Installation on the VM ======"
+    echo "Revert to snapshot defaultState"
+    kubectl get nodes -o wide --no-headers=true | awk '{print $1}' | while read HOSTNAME; do
+        govc snapshot.revert -k=true -vm=${HOSTNAME} defaultState
```

Comment:
After reverting snapshot the vm could be inaccessible for seconds. You can refer to https://github.com/antrea-io/antrea/blob/76ee297ee6ea8b10113cabba7b1f6401b4069a14/ci/jenkins/test.sh#L354 to check whether the vm is available.

Reply:
in our case we are just ending the test after reverting the snapshots so i don't think this wait is required here in our case.

Reply:
Why reverting snapshot after ending test? I think there is a potential risk that if testbed is modified by some reasons, the first round test will be run with different setup. It's one of the reasons why we reverted snapshot before running test in previous cases.

Reply:
okay